src/RequirementChecker/AppRequirementsFactory.php

```diff
@@ -0,0 +1,204 @@
+
+        foreach ($extensionRequirements as $extension => $packages) {
+            foreach ($packages as $package) {
+                if (self::SELF_PACKAGE === $package) {
+                    $message = sprintf(
+                        'The application requires the extension "%s". Enable it or install a polyfill.',
+                        $extension
+                    );
+                    $helpMessage = sprintf(
+                        'The application requires the extension "%s".',
+                        $extension
+                    );
+                } else {
+                    $message = sprintf(
+                        'The package "%s" requires the extension "%s". Enable it or install a polyfill.',
+                        $package,
+                        $extension
+                    );
+                    $helpMessage = sprintf(
+                        'The package "%s" requires the extension "%s".',
+                        $package,
+                        $extension
+                    );
+                }
+
+                $requirements->addRequirement(
+                    self::generateExtensionCheckStatement($extension),
+                    $message,
+                    $helpMessage
+                );
+            }
+        }
+    }
+
+    private static function exportRequirementsIntoConfig(RequirementCollection $requirements): array
+    {
+        return array_map(
+            function (Requirement $requirement): array {
+                return [
+                    $requirement->getIsFullfilledChecker(),
+                    $requirement->getTestMessage(),
+                    $requirement->getHelpText(),
+                ];
+            },
+            $requirements->getRequirements()
+        );
+    }
+
+    /**
+     * Collects the extension required. It also accounts for the polyfills, i.e. if the polyfill `symfony/polyfill-mbstring` is provided
+     * then the extension `ext-mbstring` will not be required.
+     *
+     * @param array $composerLockContents
+     *
+     * @return array Associative array containing the list of extensions required
+     */
+    private static function collectExtensionRequirements(array $composerLockContents): array
+    {
+        $requirements = [];
+        $polyfills = [];
+
+        $platform = $composerLockContents['platform'] ?? [];
+
+        foreach ($platform as $package => $constraint) {
+            if (preg_match('/^ext-(?<extension>.+)$/', $package, $matches)) {
+                $extension = $matches['extension'];
+
+                $requirements[$extension] = [self::SELF_PACKAGE];
+            }
+        }
+
+        $packages = $composerLockContents['packages'] ?? [];
+
+        foreach ($packages as $packageInfo) {
+            $packageRequire = $packageInfo['require'] ?? [];
+
+            if (1 === preg_match('/symfony\/polyfill-(?<extension>.+)/', $packageInfo['name'], $matches)) {
```

Comment:
I could be interesting to add the ability to have another polyfill format: the current implementation only allow polyfill from Symfony.

There are other polyfill like [mcrypt_compat](https://github.com/phpseclib/mcrypt_compat) or [random_compat](https://github.com/paragonie/random_compat) (with the name containing `FEATURE_compat`)